Type annotations are frequently ignored in --checkJs

**TypeScript Version:**  2.3.2

**Code**

```js
class Wrapper {
  constructor() {
    /** @type {boolean} */
    this.isThing = false;
    this.isThing = !this.isThing;
  }
}
```

**Expected behavior:**

This passes.

**Actual behavior:**

Every once in a while:

> test.js(4,5): error TS7022: 'isThing' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.

It's very inconsistent whether or not this error (and other similar errors) appears. It seems like being part of a larger project makes it much more likely.

Since this bug is much easier to consistently reproduce on larger codebases, I've pushed a commit that reproduces the bug in the context I hit upon it:

https://github.com/Zarel/Pokemon-Showdown/tree/e34c77930a7eaae05a28eb68cbc392c777b8c590

If you checkout this commit and run `tsc`, you'll get three errors which I believe are related:

> sim/dex-data.js(101,3): error TS7022: 'exists' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
>
> sim/dex.js(87,7): error TS2322: Type 'string[]' is not assignable to type '("Pokedex" | "FormatsData" | "Learnsets" | "Movedex" | "Statuses" | "TypeChart" | "Scripts" | "It...'.
>
> Type 'string' is not assignable to type '"Pokedex" | "FormatsData" | "Learnsets" | "Movedex" | "Statuses" | "TypeChart" | "Scripts" | "Ite...'.
>
> sim/dex.js(1476,33): error TS2531: Object is possibly 'null'.

 (All three of them disappear when I copy/paste the relevant code into a new file to try to create a simple testcase, and all three are incorrect.)

Here's a screenshot of it failing in Visual Studio Code:

![image](https://cloud.githubusercontent.com/assets/551184/25831066/f565c7f8-3426-11e7-8d0f-122725a412e7.png)
![image](https://cloud.githubusercontent.com/assets/551184/25831108/55025898-3427-11e7-989e-b10ab2bc7889.png)

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -4158,6 +4158,7 @@ namespace ts {
             const types: Type[] = [];
             let definedInConstructor = false;
             let definedInMethod = false;
+            let jsDocType: Type;
             for (const declaration of symbol.declarations) {
                 const expression = declaration.kind === SyntaxKind.BinaryExpression ? <BinaryExpression>declaration :
                     declaration.kind === SyntaxKind.PropertyAccessExpression ? <BinaryExpression>getAncestor(declaration, SyntaxKind.BinaryExpression) :
@@ -4176,19 +4177,26 @@ namespace ts {
                     }
                 }
 
-                if (expression.flags & NodeFlags.JavaScriptFile) {
-                    // If there is a JSDoc type, use it
-                    const type = getTypeForDeclarationFromJSDocComment(expression.parent);
-                    if (type && type !== unknownType) {
-                        types.push(getWidenedType(type));
-                        continue;
+                // If there is a JSDoc type, use it
+                const type = getTypeForDeclarationFromJSDocComment(expression.parent);
+                if (type) {
+                    const declarationType = getWidenedType(type);
+                    if (!jsDocType) {
+                        jsDocType = declarationType;
+                    }
+                    else if (jsDocType !== unknownType && declarationType !== unknownType && !isTypeIdenticalTo(jsDocType, declarationType)) {
+                        const name = getNameOfDeclaration(declaration);
+                        error(name, Diagnostics.Subsequent_variable_declarations_must_have_the_same_type_Variable_0_must_be_of_type_1_but_here_has_type_2, declarationNameToString(name), typeToString(jsDocType), typeToString(declarationType));
                     }
                 }
-
-                types.push(getWidenedLiteralType(checkExpressionCached(expression.right)));
+                else if (!jsDocType) {
+                    // If we don't have an explicit JSDoc type, get the type from the expression.
+                    types.push(getWidenedLiteralType(checkExpressionCached(expression.right)));
+                }
             }
 
-            return getWidenedType(addOptionality(getUnionType(types, /*subtypeReduction*/ true), definedInMethod && !definedInConstructor));
+            const type = jsDocType || getUnionType(types, /*subtypeReduction*/ true);
+            return getWidenedType(addOptionality(type, definedInMethod && !definedInConstructor));
         }
 
         // Return the type implied by a binding pattern element. This is the type of the initializer of the element if
@@ -13229,7 +13237,7 @@ namespace ts {
                 }
                 return result;
             }
-       }
+        }
 
         function isValidSpreadType(type: Type): boolean {
             return !!(type.flags & (TypeFlags.Any | TypeFlags.Null | TypeFlags.Undefined | TypeFlags.NonPrimitive) ||

--- a/tests/baselines/reference/checkJsFiles7.symbols
+++ b/tests/baselines/reference/checkJsFiles7.symbols
@@ -1,0 +1,20 @@
+=== tests/cases/compiler/a.js ===
+class C {
+>C : Symbol(C, Decl(a.js, 0, 0))
+
+	constructor() {
+		/** @type {boolean} */
+		this.a = true;
+>this.a : Symbol(C.a, Decl(a.js, 1, 16), Decl(a.js, 3, 16))
+>this : Symbol(C, Decl(a.js, 0, 0))
+>a : Symbol(C.a, Decl(a.js, 1, 16), Decl(a.js, 3, 16))
+
+		this.a = !!this.a;
+>this.a : Symbol(C.a, Decl(a.js, 1, 16), Decl(a.js, 3, 16))
+>this : Symbol(C, Decl(a.js, 0, 0))
+>a : Symbol(C.a, Decl(a.js, 1, 16), Decl(a.js, 3, 16))
+>this.a : Symbol(C.a, Decl(a.js, 1, 16), Decl(a.js, 3, 16))
+>this : Symbol(C, Decl(a.js, 0, 0))
+>a : Symbol(C.a, Decl(a.js, 1, 16), Decl(a.js, 3, 16))
+	}
+}

--- a/tests/baselines/reference/checkJsFiles7.types
+++ b/tests/baselines/reference/checkJsFiles7.types
@@ -1,0 +1,25 @@
+=== tests/cases/compiler/a.js ===
+class C {
+>C : C
+
+	constructor() {
+		/** @type {boolean} */
+		this.a = true;
+>this.a = true : true
+>this.a : boolean
+>this : this
+>a : boolean
+>true : true
+
+		this.a = !!this.a;
+>this.a = !!this.a : boolean
+>this.a : boolean
+>this : this
+>a : boolean
+>!!this.a : boolean
+>!this.a : boolean
+>this.a : true
+>this : this
+>a : true
+	}
+}

--- a/tests/cases/compiler/checkJsFiles7.ts
+++ b/tests/cases/compiler/checkJsFiles7.ts
@@ -1,0 +1,12 @@
+// @allowJs: true
+// @checkJs: true
+// @noEmit: true
+// @noImplicitAny: true
+// @fileName: a.js
+class C {
+	constructor() {
+		/** @type {boolean} */
+		this.a = true;
+		this.a = !!this.a;
+	}
+}

